src/nyx/features/ngldm.cpp

```diff
@@ -0,0 +1,346 @@
+#include "ngldm.h"
+#include "../environment.h"
+
+NGLDMfeature::NGLDMfeature() : FeatureMethod("NGLDMfeature")
+{
+	provide_features ({
+		NGLDM_LDE,
+		NGLDM_HDE,
+		NGLDM_LGLCE,
+		NGLDM_HGLCE,
+		NGLDM_LDLGLE,
+		NGLDM_LDHGLE,
+		NGLDM_HDLGLE,
+		NGLDM_HDHGLE,
+		NGLDM_GLNU,
+		NGLDM_GLNUN,
+		NGLDM_DCNU,
+		NGLDM_DCNUN,
+		NGLDM_GLM,
+		NGLDM_GLV,
+		NGLDM_DCM,
+		NGLDM_DCV,
+		NGLDM_DCE,
+		NGLDM_DCENE });
+}
+
+void NGLDMfeature::clear_buffers()
+{
+	f_LDE =
+		f_HDE =
+		f_LGLCE =
+		f_HGLCE =
+		f_LDLGLE =
+		f_LDHGLE =
+		f_HDLGLE =
+		f_HDHGLE =
+		f_GLNU =
+		f_GLNUN =
+		f_DCNU =
+		f_DCNUN =
+		f_GLCM =
+		f_GLV =
+		f_DCM =
+		f_DCV =
+		f_DCE =
+		f_DCENE = 0;
+}
+
+template <class PixelCloud> void NGLDMfeature::gather_unique_intensities (std::vector<PixIntens> & V, PixelCloud & C, PixIntens max_inten)
+{
+	std::unordered_set<PixIntens> U;
+	PixIntens range = max_inten - 0;
+	unsigned int nGrays = theEnvironment.get_coarse_gray_depth();
+	for (Pixel2 p : C)
+	{
+		PixIntens inten_ = Nyxus::to_grayscale(p.inten, 0, range, nGrays, Environment::ibsi_compliance);
+		U.insert(inten_);
+	}
+
+	// -- Set to vector to be able to know each intensity's index
+	V.insert (V.end(), U.begin(), U.end()); //std::vector<PixIntens> V (U.begin(), U.end());
+	std::sort (V.begin(), V.end());
+}
+
+template <class Imgmatrix> void NGLDMfeature::calc_ngldm (SimpleMatrix<unsigned int> & NGLDM, Imgmatrix & I, std::vector<PixIntens> & V, PixIntens max_inten)
+{
+	// Define the neighborhood
+	struct ShiftToNeighbor
+	{
+		int dx, dy;
+	}; 
+	const static ShiftToNeighbor shifts[] =
+	{
+		{-1, 0},	// West
+		{-1, -1},	// North-West
+		{0, -1},	// North
+		{1, -1},	// North-East
+		{1, 0},		// East
+		{1, 1},		// South-East
+		{0, 1},		// South
+		{-1, 1}		// South-West
+	};
+
+	// Temps
+	PixIntens range = max_inten - 0;
+	unsigned int nGrays = theEnvironment.get_coarse_gray_depth();
+
+	// Iterate pixels having all the 8 neighbors
+	for (int y = 1; y < I.get_height() - 1; y++)
+		for (int x = 1; x < I.get_width() - 1; x++)
+		{
+			// Raw intensity of the central pixel whose 
+			PixIntens cpi = I.yx (y,x);
+
+			// Skip off-ROI pixels
+			if (cpi == 0)
+				continue;
+
+			// Binned intensity
+			PixIntens cpi_ = Nyxus::to_grayscale(cpi, 0, range, nGrays, Environment::ibsi_compliance);	// binned 'cpi'
+
+			// Get a dense index value for sparse binned intensity cpi_
+			auto iter = std::find(V.begin(), V.end(), cpi_);
+			int row = (int)(iter - V.begin());
+
+			// Analyze the neighborhood histogram of pixel (y,x)
+			for (int i = 0; i < 8; i++)
+			{
+				PixIntens npi = I.yx(y + shifts[i].dy, x + shifts[i].dx);	// neighboring pixel intensity
+				PixIntens npi_ = Nyxus::to_grayscale(npi, 0, range, nGrays, Environment::ibsi_compliance);	// binned 'npi'
+
+				bool match = cpi_ == npi_;
+
+				if (!match)
+					continue;
+
+				unsigned int& binCount = NGLDM.yx(row, i);
+				binCount += (int)match;
+			}
+		}
+}
+
+void NGLDMfeature::calculate (LR& r)

```

Comment:
similar observation here. It is not clear to me whether `r` changes inside the function. If it does not change, I would prefer a `const`.

Reply:
"r" may be changed, for example, a feature class X can create a per-ROI object and leave it in "r" for another feature class Y that depends on X. With this purpose in mind, method "calculate(.)" is declared up in abstract class "FeatureMethod", parent of "NGLDMfeature". Examples of calculate(LR & r) that modify r's state is the contour feature leaving contour pixels container for the radial distribution features calculated after the contour, the convex hull feature leaving the convex hull pixels container for caliper features calculated after the contour, etc.